partiql-lang/src/main/kotlin/org/partiql/lang/eval/PartiqlAstExtensions.kt

```diff
@@ -65,3 +65,114 @@ internal fun PartiqlAst.Expr.getStartingSourceLocationMeta(): SourceLocationMeta
+            var current = emptySet<String>()
+
+            // SELECT clauses are processed in the reverse of evaluation order, since that's how variable scoping extends.
+
+            // Note: in `SELECT e AS x ...`, free variables of `e` get into the result,
+            // but `x` *does not*, due to `x` being a 'symbol', not an 'id' in the AST.
+            // (For reference, 'x' *should not* count as either free or bound because it is not a variable,
+            // it is a spec to create an attribute in a struct.)
+            current += walkProjection(node.project, emptySet())
+
+            node.limit?.let { current += walkExpr(it, emptySet()) }
+            node.offset?.let { current += walkExpr(it, emptySet()) }
+
+            node.order?.let { current += walkOrderBy(it, emptySet()) }
+
+            node.having?.let { current += walkExpr(it, emptySet()) }
+
+            node.group?.let {
+                val freeInGroup = walkGroupBy(it, emptySet())
+                val boundByGroup =
+                    (it.groupAsAlias?.let { setOf(it.text) } ?: emptySet()) +
+                        it.keyList.keys.flatMap { it.asAlias?.let { setOf(it.text) } ?: emptySet() }
+                current = (current - boundByGroup) + freeInGroup
+            }
+
+            node.where?.let { current += walkExpr(it, emptySet()) }
+
+            node.fromLet?.let {
+                current = addFreeOfLets(it.letBindings, current)
+            }
+
+            current = addFreeOfFrom(node.from, current)
+
+            return accumulator + current
+        }
+
+        // Accounts for situations when an earlier binder is referenced in a later FROM item, such as t in 2nd item in
+        //   FROM Tbl as t, t as x
+        // Such a reference should not be counted as free.
+        private fun addFreeOfFrom(n: PartiqlAst.FromSource, accum: Set<String>): Set<String> =
+            when (n) {
+                is PartiqlAst.FromSource.Join -> {
+                    val acc = addFreeOfFrom(n.right, accum)
+                    addFreeOfFrom(n.left, acc)
+                }
+                is PartiqlAst.FromSource.Scan -> {
+                    val binders = setOfNotNull(n.asAlias?.text, n.atAlias?.text, n.byAlias?.text)
+                    (accum - binders) + walkExpr(n.expr, emptySet())
+                }
+                is PartiqlAst.FromSource.Unpivot -> {
+                    val binders = setOfNotNull(n.asAlias?.text, n.atAlias?.text, n.byAlias?.text)
+                    (accum - binders) + walkExpr(n.expr, emptySet())
+                }
+            }
+
+        // Similarly to FROM, accounts for, e.g.,  LET expr as x, 2*x+1 as z   -- 2nd x is not free.
+        private fun addFreeOfLets(ns: List<PartiqlAst.LetBinding>, accum: Set<String>): Set<String> =
+            ns.foldRight(accum) { n, acc ->
+                (acc - n.name.text) + walkExpr(n.expr, emptySet())
+            }
```

Comment:
Can you help me clarify the expected behavior when an identifier is used within the original expression and its textual format is the alias?

Consider:
```sql
SELECT a
FROM t AS t
LET t AS t
```

In the visitor, we would:
1. Evaluate PROJECT. We'd accumulate `set(a)`.
2. Evaluate LET: We'd accumulate: `(acc - n.name.text) + walkExpr(n.expr, emptySet())` == `(set(a) - t) + set(t)` == `set(a) + set(t)` == `set(a, t)`. Unsure if this is expected.
3. Evalute SCAN: `(accum - binders) + walkExpr(n.expr, emptySet())` == `(set(a, t) - set(t)) + set(t)` == `set(a) + set(t)` == `set(a, t)`. Also unsure if this is expected.

Reply:
Yes, this is as expected.  I have added a couple tests inspired by this example.

Intuitively, "free variables" of an expression are "all references to variables whose definitions are external/global w.r.t. this expression". 

For the sake of discussion, let's mark each occurrence of `t` in the example (but still think of it as the same variable):

```sql
SELECT a
FROM t1 AS t2
LET t3 AS t4
```

Then, what we want is for `freeVars` is to reflect the fact that the "externals" are `a` and `t1` -- which it does.  For the remaining, t2, t3, t4 do not enter the final free variables set: `t3` is a reference to the binder `t2`, and `t4` is a binder itself (available to be referenced in SELECT, but isn't being used there).